### DISCUSSION TITLE
Can the table of contents include JSX headings?

### DISCUSSION THREAD

**Discussion by KamilDev** [fuma-nama/fumadocs#1170] [Q&A]
How can I make the table of contents include JSX headings. For example:
```mdx
---
title: Fumadocs Test
---

# Markdown heading

<h1>JSX heading</h1>
```

I would like "JSX heading" to be included in TOC, but it isn't. In the discussion #939 the following is stated:
> The heading JSX elements won't be included in TOC.

Is there any way to have them be included?

**Comment by fuma-nama** [comment]
No, this is also considered as a way to exclude headings from TOC so it is unlikely to be changed. There’s no a significant reason that requires parsing JSX headings, you can always use JSX even inside markdown headings

**Reply by KamilDev** [reply]
@fuma-nama What would the approach be to allow templating in MDX files for similar pages. For example, every API endpoint page has the following:
```tsx
export function APIDocumentationTables({
  apiDoc
}: APIDocumentationTablesProps) {
  return (
    <>
      <h2>Request Headers</h2>
      <APIHeadersTable items={apiDoc.headers} />

      <h2>Request Body</h2>
      <APIFieldsTable items={apiDoc.request} />

      <h2>Success Response</h2>
      <APIResponseTable items={apiDoc.response} />

      <h2>Error Response</h2>
      <APIResponseTable items={apiDoc.errorResponse} />

      <h2>Response Codes</h2>
      <APIResponseCodesTable items={apiDoc.responseCodes} />
    </>
  );
}
```

I can then easily use this in MDX without duplicating code:
```mdx
<APIDocumentationTables apiDoc={apiDocs.converter} />
```

This allows in the future to, for example, rename the headings, add additional headings or tables, and have all the pages updated without manually having to modify every single MDX page.

This example is why I was wondering about the JSX headings in TOC. What approach do I take instead?

**Reply by fuma-nama** [reply]
It is not possible because we cannot access the rendered children of JSX nodes from MDX compiler, unless we sacrifice performance and try to reconstruct the components, render them into html, and find all heading tags. 

One better way is to hardcode or generate the TOC in your component, and return that TOC to parent via function props, then pass the TOC to page (which bypasses the TOC generated at compile process, using the one generated at runtime in RSC)

I’m also thinking of a option to parse headings on browser from DOM tree, but it might work bad with SSR so it is probably not valid in your case

**Reply by KamilDev** [reply]
> One better way is to hardcode or generate the TOC in your component, and return that TOC to parent via function props, then pass the TOC to page (which bypasses the TOC generated at compile process, using the one generated at runtime in RSC)

Would this still work if the component is used within other unique MDX headings, if so what's the approach? For example:
```mdx
## Some unique heading

Content here

<APIDocumentationTables apiDoc={apiDocs.converter} />

## Another unique heading

Content here
```

Another idea is having the ability to predefine a static template string, and have it automatically inserted inside the MDX file at the earliest point before MDX compiler begins. I believe it would only need to happen when the template string or MDX file has been changed. Though it's similar to "reconstruct the components", perhaps it can be performant due to the simplistic nature. Example:

```APIDocumentationTables.template.mdx```
```mdx
 ## Request Headers

<APIHeadersTable items={ARG_apiDoc.headers} className="mt-4" />

## Request Body

<APIFieldsTable items={ARG_apiDoc.request} className="mt-4" />

## Etc..
```

`some-page.mdx`
```mdx
Content here

<template:APIDocumentationTables apiDoc={apiDocs.converter}  />

More content here
```

Essentially it would just be automating copy and pasting the template into the file. The file would result to:
```mdx
Content here

 ## Request Headers

<APIHeadersTable items={apiDocs.converter.headers} className="mt-4" />

## Request Body

<APIFieldsTable items={apiDocs.converter.request} className="mt-4" />

## Etc..

More content here
```

**Reply by fuma-nama** [reply]
The problem is we must not render the components in compilation phrase, which makes it only possible to obtain it on runtime. I think I can work on some examples to bring you the ideas more clearly later

**Reply by KamilDev** [reply]
I think what I'm looking for is to use a templating engine such as [Handlebars.js](https://github.com/handlebars-lang/handlebars.js) or [LiquidJS](https://github.com/harttle/liquidjs). I would have a file watcher setup, and introduce it as precompile step, so MDX compiler only gets the final output file from the template parsing result.
